File: src/etch/examples/chat/ChatClientImpl.cs
using System;
using System.Collections.Generic;
using System.Text;

using Etch.Msg;
using Etch.Support;
using Etch.Transport;

namespace etch.examples.chat
{
    public class ChatClientImpl : ChatClient, Session
    {
        private readonly String _whoami;
        private readonly bool _snub;
        private List<String> whoisOnline = new List<string>();
        private ChatServer server;


        /// <summary>
        /// Constructs the ChatClientImpl
        /// </summary>
        /// <param name="whoami"></param>
        /// <param name="snub"></param>
        public ChatClientImpl( String whoami, bool snub )
        {
            _whoami = whoami;
            _snub = snub;
        }

        #region ChatClient Members

        public void whoIsOnline( string[] names )
        {
            foreach ( String name in names )
            {
                Console.WriteLine( "{0}: {1} online (src = whoIsOnline)\n", _whoami, name );
                whoisOnline.Add( name );
            }
        }

        public void statusChange( string name, bool? online )
        {
            if ( (bool) online )
            {
                Console.WriteLine( "{0}: {1} online (src = statusChange)\n", _whoami, name );
                whoisOnline.Add( name );
            }
            else
            {
                Console.WriteLine( " {0}: {1} offline\n", _whoami, name );
                whoisOnline.Remove( name );
            }
        }

        #endregion

        #region Chat Members

        public bool? isLoggedIn()
        {
            // server is always logged in to the client
            return true;
        }

        public void send( string who, string msg )
        {
            Console.WriteLine( "{0}: msg from {1}: {2}\n", _whoami, who, msg );
            if ( _snub )
            {
                server.send( who, String.Format( "{0}, leave me alone!", who ) );
                Console.WriteLine( "snub done" );
            }
        }

        #endregion

        /// <summary>
        /// Sets the server so that we can send messages to it. 
        /// </summary>
        /// <param name="server"></param>
        public void setServer( ChatServer server )
        {
            this.server = server;
        }


        #region Session Members

        public void SessionControl( object control, object value )
        {
            throw new Exception( "The method or operation is not implemented." );
        }

        public void SessionNotify( object eventObj )
        {
            if ( eventObj.Equals( SessionConsts.DOWN ) )
                Console.WriteLine( "(server disconnected)" );
        }

        public object SessionQuery( object query )
        {
            throw new Exception( "The method or operation is not implemented." );
        }

        #endregion
    }
}

File: src/etch/examples/chat/ChatClientImpl.java
package etch.examples.chat;

import java.util.HashSet;
import java.util.Set;

import etch.bindings.java.transport.Session;

/**
 * Test jig vs. ChatServerImpl.
 */
public class ChatClientImpl implements ChatClient, Session
{
	/**
	 * Constructs the ChatClientImpl.
	 * @param whoami
	 * @param snub
	 */
	public ChatClientImpl( String whoami, boolean snub )
	{
		this.whoami = whoami;
		this.snub = snub;
	}
	
	private final String whoami;
	
	private final boolean snub;

	public Boolean isLoggedIn()
	{
		// server is always logged in to the client.
		return true;
	}

	public void statusChange( String name, Boolean online ) 
	{
		if (online)
		{
			System.out.printf( "%s: %s online (src=statusChange)\n", whoami, name );
			whoIsOnline.add( name );
		}
		else
		{
			System.out.printf( "%s: %s offline\n", whoami, name );
			whoIsOnline.remove( name );
		}
	}

	public void whoIsOnline( String[] names )
    {
		for (String name: names)
		{
			System.out.printf( "%s: %s online (src=whoIsOnline)\n", whoami, name );
			whoIsOnline.add(  name );
		}
    }
	
	private Set<String> whoIsOnline = new HashSet<String>();

	public void send( String who, String msg ) throws Failure
    {
		System.out.printf( "%s: msg from %s: %s\n", whoami, who, msg );
		if (snub)
		{
			server.send( who, String.format( "%s, leave me alone!", who ) );
			System.out.println( "snub done" );
		}
    }

	/**
	 * Sets the server so that we can send messages to it.
	 * @param server
	 */
	public void setServer( ChatServer server)
	{
		this.server = server;
	}
	
	private ChatServer server;

	public Object sessionQuery( Object query ) throws Exception
	{
		throw new UnsupportedOperationException( "unknown query "+query );
	}

	public void sessionControl( Object control, Object value ) throws Exception
	{
		throw new UnsupportedOperationException( "unknown control "+control );
	}

	public void sessionNotify( Object event ) throws Exception
	{
		if (event == Session.DOWN)
			System.out.println( "(server disconnected)" );
	}
}
